Firestore/core/src/firebase/firestore/util/exception_apple.mm

```diff
@@ -14,34 +14,58 @@
  * limitations under the License.
  */
 
-#include "Firestore/core/src/firebase/firestore/util/exception.h"
+#include "Firestore/core/src/firebase/firestore/util/exception_apple.h"
 
 #import <Foundation/Foundation.h>
 
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
 #include "Firestore/core/src/firebase/firestore/util/string_apple.h"
 
 NS_ASSUME_NONNULL_BEGIN
 
 namespace firebase {
 namespace firestore {
 namespace util {
-namespace internal {
+namespace {
 
-static NSException* MakeException(NSString* name, const std::string& message) {
-  return [[NSException alloc] initWithName:name
+NSString* ExceptionName(Exception exception) {
+  switch (exception) {
+    case Exception::AssertionFailure:
+      return @"FIRESTORE INTERNAL ASSERTION FAILED";
+    case Exception::IllegalState:
+      return @"FIRIllegalStateException";
+    case Exception::InvalidArgument:
+      return @"FIRInvalidArgumentException";
+  }
+  UNREACHABLE();
+}
+
+NSException* MakeException(Exception type, const std::string& message) {
+  return [[NSException alloc] initWithName:ExceptionName(type)
                                     reason:MakeNSString(message)
                                   userInfo:nil];
 }
 
-[[noreturn]] void ThrowIllegalState(const std::string& message) {
-  @throw MakeException(@"FIRIllegalStateException", message);  // NOLINT
-}
+}  // namespace
 
-[[noreturn]] void ThrowInvalidArgument(const std::string& message) {
-  @throw MakeException(@"FIRInvalidArgumentException", message);  // NOLINT
+ABSL_ATTRIBUTE_NORETURN void ObjcThrowHandler(Exception type,
+                                              const char* file,
+                                              const char* func,
+                                              int line,
+                                              const std::string& message) {
+  if (type == Exception::AssertionFailure) {
+    [[NSAssertionHandler currentHandler]
+        handleFailureInFunction:MakeNSString(func)
+                           file:MakeNSString(file)
+                     lineNumber:line
+                    description:@"%@: %s", ExceptionName(type),
+                                message.c_str()];
+    abort();
```

Comment:
Nit: include `<cstdlib>`? Also, what's the rationale to prefer `abort` to `std::terminate`?

Reply:
No good reason. I think it's just that when I wrote this it was `abort` and Rich wrote the `FailureHandler` using `std::terminate`. I've made them both `std::terminate` (and included `<exception>`, which defines it).